src/main/java/org/mockito/plugins/DoNotMockEnforcer.java

```diff
@@ -10,14 +10,32 @@
 public interface DoNotMockEnforcer {
 
     /**
-     * If this type is allowed to be mocked. Return an empty optional if the enforcer allows
+     * If this type is allowed to be mocked. Return {@code null} if the enforcer allows
      * this type to be mocked. Return a message if there is a reason this type can not be mocked.
-     *
+     * <p>
      * Note that Mockito performs traversal of the type hierarchy. Implementations of this class
      * should therefore not perform type traversal themselves but rely on Mockito.
      *
      * @param type The type to check
-     * @return Optional message if this type can not be mocked, or an empty optional if type can be mocked
+     * @return Optional message if this type can not be mocked, or {@code null} otherwise
+     *
+     * @see #checkTypeForDoNotMockStaticViolation(Class)
      */
     String checkTypeForDoNotMockViolation(Class<?> type);
+
+    /**
+     * If this type is allowed to be <em>statically</em> mocked. Return {@code null} if the enforcer
+     * allows this type to be mocked. Return a message if there is a reason this type can not be mocked.
+     * <p>
+     * Note that Mockito performs traversal of the type hierarchy. Implementations of this class
+     * should therefore not perform type traversal themselves but rely on Mockito.
+     *
+     * @param type The type to check
+     * @return Optional message if this type can not be <em>statically</em> mocked, or {@code null} otherwise
+     *
+     * @see #checkTypeForDoNotMockViolation(Class)
+     */
+    default String checkTypeForDoNotMockStaticViolation(Class<?> type) {
```

Comment:
Rather than adding a new method for each type of mock we can have (we also have constructor mocks), I think it's better to introduce an overload of `checkTypeForDoNotMockViolation` which takes in a `MockType` enum. E.g.

```
default String checkTypeForDoNotMockViolation(Class<?> clazz, MockType mockType) {
    return checkTypeForDoNotMockViolation(clazz);
}
```

Then, we can refactor `MockitoCore` to pass in the `MockType` enum as well, avoiding the boolean-parameters.

Reply:
This was my original idea, but I found it difficult to not create a possible confusion when someone overrides both methods, since `checkTypeForDoNotMockViolation` would still be required to be overridden by default.

If we're creating some sort of method like that, I'd rather wrap the parameters inside an object (e.g. something like `MockCreationSettings`) instead of having to create a new overload each time we want to expand it (however unlikely that may be).

Reply:
Works for me. Great idea, btw. Sounds like a good path forward.

One thing regarding the navigation of the class hierarchy that we make, while we are talking about this class.

For regular mocking, I think it kinda makes sense to have the navigation being done externally (currently in `MockitoCore`) and just checking the enforcer for each type, but for static mocks it makes slightly less sense to just force it to be that way, because a static mock does not have a usual "is a" relationship with the parent classes the same way a regular mock does.

We still want to *default* to it, since a class annotated with `DoNotMock` should still default to not allow static mocks too, I believe, but I think we should make it possible to override such navigation, at the very least for static mocks. I can put something in the container object that makes it easier to navigate the hierarchy the way we currently do it (think a lazy `Iterator` or `Stream` or something).

One example is a rule I'd like to implement, that would go something like "do not statically mock JRE classes". I recently found an instance where someone mocked the `java.nio.files.Files` class and that broke JAR loading as a whole, but in general I think _*statically*_ mocking JRE classes is likely a Bad Idea :tm:, and implementing such a rule becomes tricky because, for example, I wouldn't like to allow someone to statically mock `java.lang.Object`, but if I don't have the class hierarchy information available and I'm unable to override the navigation, it becomes "impossible" (well, technically possible, but hacky as hell) to create such a rule, because it would inevitably come to `java.lang.Object` at some point. 

That goes for any instance where I'm ok with someone statically mocking their own extended type, but not the JRE type they extend, e.g. it's ok for them to statically mock their own List subtype, but not with them mocking `java.util.List` itself, for example.

Reply:
So the bottom line is, I'd like to expand a bit further on your idea so I can also control the navigation of the class hierarchy, if I so desire; while still having a very sane default for most usages.

Reply:
Yes that's a good point. Can we move the navigation into `DoNotMockEnforcer`? E.g. we call `checkTypeForDoNotMockViolation` from `MockitoCore` from the top-level type and then `DoNotMockEnforcer` implements the type-traversal there. In fact, while I think about it, that makes like a solid interface we should keep. Then users can do more granular type traversal on the root interface, but if they only want to check types, they implement the "arguably easier" interface `DoNotMockEnforcer` that we already ship. Then we don't need to mark it as deprecated, as it is still a convenience

Reply:
Using `MockCreationSettings` poses a problem for caching, though. Because the user would be able to use any data inside of it, it's hard to know exactly when it's safe to cache it or not. I'd lean towards either:

1. create a container type that we are in control of and can easily ensure can be used as the key for the cache; or
2. not cache at all and not be responsible for a drop in performance from performing such evaluations; or
3. only cache for the existing `DoNotMockEnforcer` but not the new `DoNotMockForTypeEnforcer`; which will cover most use-cases anyway.

Any preferred solution and/or other ideas, @TimvdLippe?

Reply:
I think for the default type traversal in `DoNotMockEnforcer` caching is a must. With the parameter object, you don't necessarily need caching. So let's not do caching there by default indeed. Therefore, I would go with option 3.

Reply:
@TimvdLippe I tried to come up with a plugin structure that kept the old `DoNotMockEnforcer` and accounted for a new `DoNotMockForTypeEnforcer` and I think it's overall more confusing than before. I will extend the interface and the default class in a way that does not make it so much worse but still give people the control they might want.

Reply:
@TimvdLippe so, it's not final yet, but I pushed a restructured version of the idea number 3. Let me know if you think there are any red flags in your opinion.

The caching part looks a bit hacky, but it's not accessible from outside the implementation. I felt moving the default implementation of the existing method into the interface made sense, given the `DefaultDoNotMockEnforcer` would have part of the implementation and the other method would be a default anyway, so this kinda centralizes everything you need to know about the class in a single place, but I'll concede it doesn't look amazingly beautiful 😅